File: src/camera.js
// Camera with deadzone follow + world bounds clamp.
export class Camera {
  constructor(viewW, viewH) {
    this.viewW = viewW;
    this.viewH = viewH;
    this.x = 0;
    this.y = 0;
    this.deadzone = { w: 40, h: 24 };
  }

  snapTo(cx, cy) {
    this.x = Math.round(cx - this.viewW / 2);
    this.y = Math.round(cy - this.viewH / 2);
  }

  follow(cx, cy, worldW, worldH) {
    const dz = this.deadzone;
    const left = this.x + (this.viewW - dz.w) / 2;
    const right = left + dz.w;
    const top = this.y + (this.viewH - dz.h) / 2;
    const bot = top + dz.h;

    if (cx < left) this.x -= left - cx;
    else if (cx > right) this.x += cx - right;
    if (cy < top) this.y -= top - cy;
    else if (cy > bot) this.y += cy - bot;

    if (worldW > this.viewW) this.x = clamp(this.x, 0, worldW - this.viewW);
    else this.x = (worldW - this.viewW) / 2;
    if (worldH > this.viewH) this.y = clamp(this.y, 0, worldH - this.viewH);
    else this.y = (worldH - this.viewH) / 2;
  }
}

function clamp(v, lo, hi) { return v < lo ? lo : v > hi ? hi : v; }

File: python/camera.py
class Camera:
    def __init__(self, view_w, view_h):
        self.view_w = view_w
        self.view_h = view_h
        self.x = 0
        self.y = 0
        self.dz_w = 40
        self.dz_h = 24

    def snap_to(self, cx, cy):
        self.x = round(cx - self.view_w / 2)
        self.y = round(cy - self.view_h / 2)

    def follow(self, cx, cy, world_w, world_h):
        left = self.x + (self.view_w - self.dz_w) / 2
        right = left + self.dz_w
        top = self.y + (self.view_h - self.dz_h) / 2
        bot = top + self.dz_h

        if cx < left:
            self.x -= left - cx
        elif cx > right:
            self.x += cx - right
        if cy < top:
            self.y -= top - cy
        elif cy > bot:
            self.y += cy - bot

        if world_w > self.view_w:
            self.x = max(0, min(self.x, world_w - self.view_w))
        else:
            self.x = (world_w - self.view_w) / 2
        if world_h > self.view_h:
            self.y = max(0, min(self.y, world_h - self.view_h))
        else:
            self.y = (world_h - self.view_h) / 2
